Sources/CartonCLI/Commands/TestRunners/BrowserTestRunner.swift

```diff
@@ -12,39 +12,165 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+import AsyncHTTPClient
 import CartonHelpers
 import CartonKit
 import Foundation
+import NIOCore
+import NIOPosix
 import PackageModel
 import TSCBasic
+import WebDriverClient
 
 private enum Constants {
   static let entrypoint = Entrypoint(fileName: "test.js", sha256: testEntrypointSHA256)
 }
 
+enum BrowserTestRunnerError: Error, CustomStringConvertible {
+  case invalidRemoteURL(String)
+  case failedToFindDriver
+
+  var description: String {
+    switch self {
+    case let .invalidRemoteURL(url): return "Invalid remote URL: \(url)"
+    case .failedToFindDriver:
+      return """
+      Failed to find WebDriver executable or remote URL to a running driver process.
+      Please make sure that you satisfied one of the followings (smaller item number lower priority):
+      1. `chromedriver`, `geckodriver`, `safaridriver`, or `msedgedriver` has been installed in `PATH`
+      2. Set `WEBDRIVER_PATH` with the path to your WebDriver executable.
+      3. Set `WEBDRIVER_REMOTE_URL` with the address of remote WebDriver like `WEBDRIVER_REMOTE_URL=http://localhost:9515`.
+      """
+    }
+  }
+}
+
 struct BrowserTestRunner: TestRunner {
   let testFilePath: AbsolutePath
   let host: String
   let port: Int
+  let headless: Bool
   let manifest: Manifest
   let terminal: InteractiveWriter
+  let eventLoopGroup = MultiThreadedEventLoopGroup(numberOfThreads: 1)
+  let httpClient: HTTPClient
+
+  internal init(
```

Comment:
nit: the convention is not to use `internal` as it's the implicit default
```suggestion
  init(
```